proposals/nnnn-custom-main-and-global-executors.md

```diff
@@ -0,0 +1,873 @@
+
+  /// Run the executor's run loop until a condition is satisfied.
+  ///
+  /// Not every `RunLoopExecutor` will support this method; you must not call
+  /// it unless you *know* that it is supported.  The default implementation
+  /// generates a fatal error.
+  ///
+  /// Parameters:
+  ///
+  /// - until condition: A closure that returns `true` if the run loop should
+  ///                    stop.
+  func run(until condition: () -> Bool) throws
+
+  /// Signal to the run loop to stop running and return.
+  ///
+  /// This method may be called from the same thread that is in the `run()`
+  /// method, or from some other thread.  It will not wait for the run loop to
+  /// stop; calling this method simply signals that the run loop *should*, as
+  /// soon as is practicable, stop the innermost `run()` invocation and make
+  /// that `run()` invocation return.
+  func stop()
+}
+```
+
+We will also add a protocol for the main actor's executor:
+
+```swift
+protocol MainExecutor: RunLoopExecutor & SerialExecutor {
+}
+```
+
+This cannot be a typealias because those will not work for Embedded Swift.
+
+We will then expose properties on `MainActor` and `Task` to allow
+users to query the executors:
+
+```swift
+extension MainActor {
+  /// The main executor, which is started implicitly by the `async main`
+  /// entry point and owns the "main" thread.
+  public static var executor: any MainExecutor { get }
+}
+
+extension Task {
+  /// The default or global executor, which is the default place in which
+  /// we run tasks.
+  public static var defaultExecutor: any TaskExecutor { get }
+}
+```
+
+There will also be an `ExecutorFactory` protocol, which is used to set
+the default executors:
+
+```swift
+/// An ExecutorFactory is used to create the default main and task
+/// executors.
+public protocol ExecutorFactory {
+  /// Constructs and returns the main executor, which is started implicitly
+  /// by the `async main` entry point and owns the "main" thread.
+  static var mainExecutor: any MainExecutor { get }
+
+  /// Constructs and returns the default or global executor, which is the
+  /// default place in which we run tasks.
+  static var defaultExecutor: any TaskExecutor { get }
+}
+
+```
+
+along with a default implementation of `ExecutorFactory` called
+`PlatformExecutorFactory` that sets the default executors for the
+current platform.
+
+Additionally, `Task` will expose a new `currentExecutor` property, as
+well as properties for the `preferredExecutor` and the
+`currentSchedulableExecutor`:
+
+```swift
+extension Task {
+  /// Get the current executor; this is the executor that the currently
+  /// executing task is executing on.
+  ///
+  /// This will return, in order of preference:
+  ///
+  ///   1. The custom executor associated with an `Actor` on which we are
+  ///      currently running, or
+  ///   2. The preferred executor for the currently executing `Task`, or
+  ///   3. The task executor for the current thread
```

Comment:
I think this should be "for the current task" instead of "current thread"

Reply:
I'm not sure there's a practical difference there; the task executors are tracked using thread local data.

Reply:
It's this code here that does the tracking: https://github.com/swiftlang/swift/blob/5a6a14785075436b8bdb2c423cd47df4945d3ec8/stdlib/public/Concurrency/Actor.cpp#L114